docs/html/reference/build-system/index.md

```diff
@@ -0,0 +1,108 @@
+(build-interface)=
+
+# Build System Interface
+
+When dealing with installable source distributions of a package, pip does not
+directly handle the build process for the package. This responsibility is
+delegated to "build backends" -- also known as "build systems". This means
+that pip needs an interface, to interact with these build backends.
+
+There are two main interfaces that pip uses for these interactions:
+
+```{toctree}
+:hidden:
+
+pyproject-toml
+setup-py
+```
+
+<!-- prettier-ignore-start -->
+[`pyproject.toml` based](pyproject-toml)
+: Standards-backed interface, that has build isolation and explicit declaration
+  of build dependencies.
+
+[`setup.py` based](setup-py)
+: Legacy interface, that we're working to migrate users away from. Has no build
+  isolation and no good mechanisms to declare build dependencies.
+<!-- prettier-ignore-end -->
+
+Details on the individual interfaces can be found on their dedicated pages,
+linked above. This document covers the nuances around which build system
+interface pip will use for a project, as well as details that apply all
+the build system interfaces that pip may use.
+
+## Determining which build system interface is used
+
+Currently, pip uses the `pyproject.toml` based build system interface, if a
+`pyproject.toml` file exists. If not, the legacy build system interface is used.
+The intention is to switch to using the `pyproject.toml` build system interface
+unconditionally and to drop support for the legacy build system interface at
+some point in the future.
+
+When performing a build, pip will mention which build system interface it is
+using. Typically, this will take the form of a message like:
+
+```none
+Building wheel for pip (PEP 517)... done
+```
+
+```none
+Building wheel for pip (setup.py)... done
+```
+
+Here, "PEP 517" refers to `pyproject.toml` based builds and "setup.py" refers
+to `setup.py` based builds.
+
+## Controlling which build system interface is used
+
+It is possible to control which build system interface is used by pip, using
+the [`--use-pep517`](install_--use-pep517) / `--no-use-pep517` flags (and
+corresponding environment variable: `PIP_USE_PEP517`).
+
+(controlling-setup_requires)=
+
+## Controlling `setup_requires`
+
+```{hint}
+This is only relevant for projects that use setuptools as the build backend,
+and use the `setup_requires` keyword argument in their setup.py file.
+```
```

Comment:
> If you have `pyproject.toml`, then you should use `build_system.requires`. But if you have to use `setup_requires` for some reason, then make sure you specify `setuptools >= 52.0`.

This doesn't really apply since this section is aimed at end-users who'll install from those packages, rather than the package authors. Do you think it is worthwhile to just straight-up say that and add a line with advice for package authors?

> If you are forced to use `setup.py` and you need `setup_requires`, be aware that you will have problems, as it does not work correctly (mention `easy_install` here). For correct behaviour you must manually install the setup requirements, and _not_ let setuptools do so automatically.

Pretty much, and the current phrasing for saying this in the PR is:

```
Usually, the best solution for dealing with packages with setup_requires is to install the packages listed in setup_requires beforehand, using a prior pip install command. This is because there is no way to control how these dependencies are located by easy_install, or how setuptools will invoke pip using pip’s command line options -- which makes it tricky to get things working appropriately.
```

(made it a code block, otherwise it looks like I'm quoting you)

> `setup.py` build - always uses easy_install. Even with latest setuptools? I don't know how else it could support `setup_requires` in a non-PEP517 build, so I assume so...

No, it calls `pip install <thing> {some-options}` instead of `easy_install <thing> {some-options}` since 52.0 -- but I'm not a 100% sure what the options are, but I'm sure that it won't be able to pass through CLI options that were passed into pip -- which is why it's worth calling out.

> `pyproject.toml`, new setuptools (`>= 52.0`). Exposes `setup_requires` via the PEP 517 "requirements to build wheel" hook, I believe.

Yes, but this applies to all setuptools versions with PEP 517 support, IIUC.

> `pyproject.toml`, old setuptools (`< 52.0`). Uses `easy_install`, workarounds below apply.

No, this is not the case. See line above.

> Overall, I think this is a messy situation and we don't want to get sucked into explaining the history of setuptools here. But I do think we need to be careful to at a minimum give accurate advice.

Yup yup! I am trying to treat that exact line -- and based on your response, I don't think I did a good-enough job! I'll try to see how I can rephrase this... :)

Reply:
> Do you think it is worthwhile to just straight-up say that and add a line with advice for package authors?

Maybe the "hint" callout at the top should explicitly say "This advice is for people trying to install projects that are still using `setup_requires` in their `setup.py`. Project authors should switch to using `pyproject.toml` to fix the issue directly."

> No, it calls pip install <thing> {some-options} instead of easy_install <thing> {some-options} since 52.0

Eww. It's better than `easy_install`, but as you say it's still bad, because of pip options not getting passed through. But that does explain why you explicitly called out version 52.0 of setuptools.

To be honest, I don't think it's worth getting into the gory details of how setuptools support has changed over time. I'd just keep it simple and go with the basic message "if a project you're installing is not using pyproject.toml, but *is* using `setup_requires`, install the build dependencies from `setup_requires` manually, before invoking pip. Newer versions of setuptools are better at automatically installing dependencies from `setup_requires`, but there are still problems and a correct install cannot be guaranteed.

>  I'll try to see how I can rephrase this... :)

Less is more, IMO. Stick to the basic recommendation, and don't get sidetracked with alternatives.

Reply:
Alright, I redid this whole thing. I think this is the only bit where I'd like you to take a look again.